### PR TITLE
perform_step! refactor for Rosenbrock

### PR DESCRIPTION
**Pull request by ParamThakkar123**
## Checklist

- [X] Appropriate tests were added
- [X] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [X] Any new documentation only uses public API
  
## Additional context

#233

### CODE REVIEW

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl [hunk 1]
```diff
@@ -495,25 +495,50 @@ end
...
+        uidx = eachindex(integrator.uprev)
+        sizeu = size(u)
+        mass_matrix = integrator.f.mass_matrix
+        utilde = du
+    end
+    if cache isa Rosenbrock33Cache
+        mass_matrix = integrator.f.mass_matrix
+        sizeu = size(u)
+        utilde = du
+    end
```
Comment (oscardssmith) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:460]:
```suggestion
    sizeu = size(u)
    mass_matrix = integrator.f.mass_matrix
    utilde = du
    if cache isa Rosenbrock34Cache
        uidx = eachindex(integrator.uprev)
    end
```

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl [hunk 2]
```diff
@@ -573,18 +610,44 @@ end
...
     @.. broadcast=false u=uprev + b1 * k1 + b2 * k2 + b3 * k3
-
     step_limiter!(u, integrator, p, t + dt)
 
-    f(fsallast, u, p, t + dt)
+    if cache isa Rosenbrock33Cache
+        f(fsallast, u, p, t + dt)
+    end
+
+    if cache isa Rosenbrock34Cache
```
Comment (oscardssmith) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:572]:
```suggestion
    elseif cache isa Rosenbrock34Cache
```

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl [hunk 3]
```diff
@@ -495,25 +495,50 @@ end
...
+        dtC41 = C41 / dt
+        dtC42 = C42 / dt
+        dtC43 = C43 / dt
+
+        dtd1 = dt * d1
+        dtd2 = dt * d2
+        dtd3 = dt * d3
+        dtd4 = dt * d4
+        dtgamma = dt * gamma
+    end
```
Comment (oscardssmith) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:487]:
```suggestion
    dtC21 = C21 / dt
    dtC31 = C31 / dt
    dtC32 = C32 / dt
    dtd1 = dt * d1
    dtd2 = dt * d2
    dtd3 = dt * d3
    dtgamma = dt * gamma

    if cache isa Rosenbrock34Cache
        dtC41 = C41 / dt
        dtC42 = C42 / dt
        dtC43 = C43 / dt

        dtd4 = dt * d4
    end
```

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl [hunk 4]
```diff
@@ -235,6 +256,23 @@ end
         end
         integrator.stats.nsolve += 1
 
+        if cache isa Rosenbrock32ConstantCache
```
Comment (oscardssmith) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:319]: this is wrong. For Rosenbrock32 the 3rd stage happens without adaptivity

Comment (ParamThakkar123) [reply lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:319]: Can you please suggest a change to make here?? I'll implement that accordingly

Comment (ParamThakkar123) [reply lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:320]: Made the changes

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl [hunk 5]
```diff
@@ -72,8 +77,15 @@ end
...
-        copyto!(tmp, k₁)
+    if cache isa Rosenbrock23Cache
+        if mass_matrix === I
+            copyto!(tmp, k₁)
+        end
+    end
+    if cache isa Rosenbrock32Cache
+        if mass_matrix === I
+            tmp .= k₁
+        end
```
Comment (oscardssmith) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:88]: these are the same (conceptually), both should use `copyto!(tmp, k₁)`

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl [hunk 6]
```diff
@@ -90,7 +102,49 @@ end
     @.. broadcast=false k₂+=k₁
     @.. broadcast=false u=uprev + dt * k₂
     stage_limiter!(u, integrator, p, t + dt)
-    step_limiter!(u, integrator, p, t + dt)
+
+    if cache isa Rosenbrock23Cache
+        step_limiter!(u, integrator, p, t + dt)
+    end
```
Comment (oscardssmith) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:108]: I think this should be running on both integrators. I think it was just missed for Rosenbrock32 because people don't care about it.

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl [hunk 7]
```diff
@@ -1734,7 +1494,7 @@ function initialize!(integrator, cache::Rosenbrock5Cache)
     integrator.k[3] = dense3
 end
 
-@muladd function perform_step!(integrator, cache::Rosenbrock5Cache, repeat_step = false)
+@muladd function perform_step!(integrator, cache::Union{Rosenbrock5Cache, Rosenbrock4Cache}, repeat_step = false)
```
[the remark this answers is not in the document]
Comment (ParamThakkar123) [reply lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:1497]: What changes are needed there exactly ??

**Comment by ParamThakkar123** [comment]
Or are there certain methods which should be kept together ??

**Comment by ChrisRackauckas** [comment]
You'll want to combine all of the caches too, just using arrays of arrays in the caches instead of different variables for each k for example

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl (hunk 7 above)
[the remark this answers is not in the document]
Comment (oscardssmith) [reply lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:1497]: Rosenbrock5Cache should have an extra stage that needs to be computed.

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas I think the recent merge caused an error here

**Comment by ChrisRackauckas** [comment]
what do you mean?

**Comment by ParamThakkar123** [comment]
I ran the tests some time ago and all the errors pointed to OrdinaryDiffEqRosenbrock.jl even though I made no changes there

**Comment by ParamThakkar123** [comment]
image

**Comment by ChrisRackauckas** [comment]
did you rebase?

**Comment by ParamThakkar123** [comment]
no

**Comment by ParamThakkar123** [comment]
I will do that

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas Can you please review the changes to see if whatever I have done so far is correct and on track ?? If any changes are needed I will make those.

**Comment by ChrisRackauckas** [comment]
This is still all written out. It should turn into loops over arrays.

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas I made a recent change to Rodas23WCache and Rodas3PCache. Is this how it should be done ??

**Comment by ChrisRackauckas** [comment]
This isn't quite right. See ExplicitRK

**Comment by ParamThakkar123** [comment]
> This isn't quite right. See ExplicitRK

I went through the explicitrk code and understood how this should be done. I will make the changes

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl [hunk 8]
```diff
@@ -629,53 +618,35 @@ end
...
-    u = uprev + a31 * k1 + a32 * k2
-    du = f(u, p, t + c3 * dt)
-    integrator.stats.nf += 1
 
-    if mass_matrix === I
-        linsolve_tmp = du + dtd3 * dT + dtC31 * k1 + dtC32 * k2
-    else
-        linsolve_tmp = du + dtd3 * dT + mass_matrix * (dtC31 * k1 + dtC32 * k2)
-    end
+    for i in 2:eachindex(A)
```
Comment (oscardssmith) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:626]: I don't think this is right since `eachindex` of a Matrix will loop over the scalars. I think you want `for i in 2:size(A, 2)`

Comment (oscardssmith) [reply lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:626]: you also probably want an `@assert` that `size(A,1)==size(A,2)==length(k)` so the compiler can know that all the sizes match up.

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl [hunk 9]
```diff
@@ -598,28 +598,17 @@ end
 
 ################################################################################
 
-@muladd function perform_step!(integrator, cache::Rosenbrock34ConstantCache,
-        repeat_step = false)
+@muladd function perform_step!(integrator, cache::Rosenbrock34ConstantCache, repeat_step = false)
```
Comment (oscardssmith) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:601]: This should just be `RosenbrockConstantCache` (with renaming/combining all the caches in https://github.com/SciML/OrdinaryDiffEq.jl/blob/master/lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl

Comment (ParamThakkar123) [reply lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:601]: Should I combine all those caches first ?? @oscardssmith

Comment (oscardssmith) [reply lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl]: yeah

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas @oscardssmith if we change all the perform_step! implementations to have just RosenbrockCache and RosenbrockConstantCache then would we need the others ?? Or Just one perform_step! for RosenbrockCache and one perform_step! for RosenbrockConstantCache would suffice ??

**Comment by oscardssmith** [comment]
> Or Just one perform_step! for RosenbrockCache and one perform_step! for RosenbrockConstantCache would suffice ??

Exactly. That's the point of the refactor.

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas @oscardssmith Can you please review the changes made ??

**Comment by ParamThakkar123** [comment]
all caches have now been combined into RosenbrockConstantCache and RosenbrockCache

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_interpolants.jl [hunk 10]
```diff
@@ -39,42 +34,37 @@ end
 end
 
 @muladd function _ode_interpolant(Θ, dt, y₀, y₁, k,
-        cache::Union{Rosenbrock23ConstantCache,
-            Rosenbrock32ConstantCache}, idxs::Nothing,
+        cache::Union{RosenbrockConstantCache}, idxs::Nothing,
```
Comment (oscardssmith) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_interpolants.jl:37]: remove all the `Union`s where there is only 1 element (style not function)

**Comment by oscardssmith** [comment]
You also are going to need to change the stiff_addsteps.jl file.

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl [hunk 11]
```diff
@@ -213,57 +177,14 @@ function alg_cache(alg::Rosenbrock32, u, rate_prototype, ::Type{uEltypeNoUnits},
...
-    linsolve::F
-    jac_config::JCType
-    grad_config::GCType
-    reltol::RTolType
-    alg::A
-    step_limiter!::StepLimiter
-    stage_limiter!::StageLimiter
-end
-
 function alg_cache(alg::ROS3P, u, rate_prototype, ::Type{uEltypeNoUnits},
```
Comment (oscardssmith) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl:188]: you also need to update all of the `alg_cache` functions in this file

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_interpolants.jl (hunk 10 above)
Comment (ParamThakkar123) [reply lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_interpolants.jl:37]: ok. got it.

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl [hunk 12]
```diff
@@ -218,7 +218,7 @@ function alg_cache(alg::ROS3P, u, rate_prototype, ::Type{uEltypeNoUnits},
         assumptions = LinearSolve.OperatorAssumptions(true))
     grad_config = build_grad_config(alg, f, tf, du1, t)
     jac_config = build_jac_config(alg, f, uf, du1, uprev, u, tmp, du2)
-    Rosenbrock33Cache(u, uprev, du, du1, du2, k1, k2, k3, k4,
+    RosenbrockCache(u, uprev, du, du1, du2, k1, k2, k3, k4,
```
Comment (oscardssmith) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl:221]: you need to combine the ks and dus

Comment (ParamThakkar123) [reply lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl:221]: Ok I will store all the ks and dus in an array an make a loop or something like that

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl [hunk 13]
```diff
@@ -20,21 +20,10 @@ end
...
-    du2::rateType
-    k1::rateType
-    k2::rateType
-    k3::rateType
-    k4::rateType
-    k5::rateType
-    k6::rateType
-    k7::rateType
-    k8::rateType
+    dus::Array{rateType, 1}  # Replacing NTuple{3, rateType} with Array{rateType, 1}
```
Comment (oscardssmith) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl:23]:
```suggestion
    dus::Vector{rateType} # Replacing NTuple{3, rateType} with Array{rateType, 1}
```

**Comment by ParamThakkar123** [comment]
ks should also be a vector ??

**Comment by ParamThakkar123** [comment]
@oscardssmith the function gen_cache_struct is causing the redefinition of RosenbrockCache and RosenbrockConstantCache. Should that be modified ??

**Comment by oscardssmith** [comment]
Comment it out for now so the code builds. I think (@ChrisRackauckas please confirm) that that code also needs to be replaced by this eventually, but it's probably a better use of your time to get the commonly used solvers working first.

**Comment by ParamThakkar123** [comment]
> Comment it out for now so the code builds. I think (@ChrisRackauckas please confirm) that that code also needs to be replaced by this eventually, but it's probably a better use of your time to get the commonly used solvers working first.

some variables are using this function

**Comment by ChrisRackauckas** [comment]
I would highly recommend taking it in steps as I mentioned. Rodas4/P and Rodas5/P are very similar just with different numbers of steps. Do those first, get tests passing, merge. Then grow it to Rodas3. Then grow it. Since Rosenbrock23 for example has different behavior in the adaptivity code, it may need a branch. But handling that at the same time as others is just making it more difficult than it needs to be.

**Comment by ParamThakkar123** [comment]
> I would highly recommend taking it in steps as I mentioned. Rodas4/P and Rodas5/P are very similar just with different numbers of steps. Do those first, get tests passing, merge. Then grow it to Rodas3. Then grow it. Since Rosenbrock23 for example has different behavior in the adaptivity code, it may need a branch. But handling that at the same time as others is just making it more difficult than it needs to be.

Okay. I got your point. I'll take the needful steps. Sorry for the previous ones

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl [hunk 14]
```diff
@@ -1734,7 +1734,7 @@ function initialize!(integrator, cache::Rosenbrock5Cache)
     integrator.k[3] = dense3
 end
 
-@muladd function perform_step!(integrator, cache::Rosenbrock5Cache, repeat_step = false)
+@muladd function perform_step!(integrator, cache::RosenbrockCache, repeat_step = false)
     @unpack t, dt, uprev, u, f, p = integrator
     @unpack du, du1, du2, k1, k2, k3, k4, k5, k6, k7, k8, dT, J, W, uf, tf, linsolve_tmp, jac_config, atmp, weight, stage_limiter!, step_limiter! = cache
     @unpack a21, a31, a32, a41, a42, a43, a51, a52, a53, a54, a61, a62, a63, a64, a65, C21, C31, C32, C41, C42, C43, C51, C52, C53, C54, C61, C62, C63, C64, C65, C71, C72, C73, C74, C75, C76, C81, C82, C83, C84, C85, C86, C87, gamma, d1, d2, d3, d4, d5, c2, c3, c4, c5 = cache.tab
```
Comment (oscardssmith) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:1740]: note that the tab also should be changed to have an `a` and `C` matrix, and a `d` and `c` vector.

Comment (ParamThakkar123) [reply lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:1740]: yeah

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl [hunk 15]
```diff
@@ -1253,55 +1235,55 @@ end
     du = f(uprev, p, t)
     integrator.stats.nf += 1
 
-    linsolve_tmp = du + dtd1 * dT
```
Comment (oscardssmith) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:1256]: this entire thing should be a `for` loop that loops over all the steps (so that it will work for arbitrary sized tableau)

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl [hunk 16]
```diff
@@ -1691,159 +1691,42 @@ end
...
-        @.. broadcast=false linsolve_tmp=du + dus[3]
-    end
+        @.. broadcast=false veck=-vecu
+        integrator.stats.nsolve += 1
 
-    linres = dolinsolve(integrator, linres.cache; b = _vec(linsolve_tmp))
-    veck7 = _vec(ks[7])
-    @.. broadcast=false veck7=-vecu
-    integrator.stats.nsolve += 1
+        @.. broadcast=false u=uprev + sum(a[i, j] * ks[j] for j in 1:i-1)
```
Comment (oscardssmith) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:1712]: This will remove the allocation
```suggestion
        @..  u=uprev
        for j in 1:i-1
            @.. u  += a[i, j] * ks[j]
       end
```

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl [hunk 17]
```diff
@@ -1691,159 +1691,42 @@ end
...
+        f(du, u, p, t + c[i] * dt)
+        integrator.stats.nf += 1
 
-    u .+= ks[7]
-    f(du, u, p, t + dt)
-    integrator.stats.nf += 1
+        if mass_matrix === I
+            @.. broadcast=false linsolve_tmp=dus[1] + dtd[i] * dT + sum(dtC[i, j] * ks[j] for j in 1:i-1)
+        else
+            @.. broadcast=false dus[2]=sum(dtC[i, j] * ks[j] for j in 1:i-1)
```
Comment (oscardssmith) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:1723]: same here
```suggestion
            @.. dus[2] = dtC[i, 1] * ks[1] 
            for j in 2:i-1
                @..  dus[2] += dtC[i, j] * ks[j] 
           end
 ```

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl [hunk 18]
```diff
@@ -658,68 +696,24 @@ end
...
     du1 = zero(rate_prototype)
     du2 = zero(rate_prototype)
+    dus = [du, du1, du2]
     k1 = zero(rate_prototype)
     k2 = zero(rate_prototype)
     k3 = zero(rate_prototype)
     k4 = zero(rate_prototype)
     k5 = zero(rate_prototype)
     k6 = zero(rate_prototype)
+    ks = [k1, k2, k3, k4, k5, k6]
```
Comment (oscardssmith) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl:716]: this can just be `zeros(typeof(rate_prototype), 6)` (not critical, just nice) and similar for `dus` and `dense`

**Comment by ParamThakkar123** [comment]
@oscardssmith can you please review this ??

**Comment by ParamThakkar123** [comment]
@oscardssmith should dense and ks also be vectors ??

**Comment by ParamThakkar123** [comment]
image

@ChrisRackauckas @oscardssmith How can I solve this error ? Looks like this error comes from the Low storage and high storage rk methods

**Comment by ParamThakkar123** [comment]
most of the tests are failing due to this one

**Comment by ParamThakkar123** [comment]
@oscardssmith @ChrisRackauckas Anything changes to be made here further ??

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl [hunk 19]
```diff
@@ -1,13 +1,46 @@
...
+struct RosenbrockConstantCache{TF, UF, Tab, JType, WType, F, AD} <: OrdinaryDiffEqConstantCache
+    tf::TF
+    uf::UF
+    tab::Tab
+    J::JType
+    W::WType
+    linsolve::F
+    autodiff::AD
+end
+
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl:43]:
```suggestion

get_fsalfirstlast(cache::RosenbrockCache) = (cache.fsalfirst, cache.fsallast)

```

**Comment by ChrisRackauckas** [comment]
It's not passing tests so it can't merge?

**Comment by ParamThakkar123** [comment]
image
this one

**Comment by ChrisRackauckas** [comment]
That's unrelated. Don't worry about it.

Your actual error is that you deleted the overloads required to make Rosenbrock23 work. Rebase this.

**Comment by ParamThakkar123** [comment]
okay

**Comment by ParamThakkar123** [comment]
image

@ChrisRackauckas How can this be dealt with ? There is some method call error here

**Comment by oscardssmith** [comment]
You need to be using https://github.com/SciML/PreallocationTools.jl somewhere to make the caches autodiff compatible.

**Comment by ChrisRackauckas** [comment]
That's done in build_jac_config.

**Comment by ParamThakkar123** [comment]
> You need to be using https://github.com/SciML/PreallocationTools.jl somewhere to make the caches autodiff compatible.

For the error I mentioned in the message above??

**Comment by ParamThakkar123** [comment]
> That's done in build_jac_config.

There's a function `build_jac_config` in the `derivative_wrappers.jl` file with AutoDiff

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas Should I be importing `build_jac_config` from `derivateive_wrappers.jl`. It doesn't seem to use it

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas @oscardssmith just needed some guidance on the current failures that occur in test. I tried to figure this out but finding it a bit hard to get where things are wrong

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas can you please review the errors here ?? I will fix those.

**Comment by ParamThakkar123** [comment]
Okay. will try to figure things out and fix them. i will try to get this done today.

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl [hunk 20]
```diff
@@ -1,13 +1,49 @@
 abstract type RosenbrockMutableCache <: OrdinaryDiffEqMutableCache end
-abstract type RosenbrockConstantCache <: OrdinaryDiffEqConstantCache end
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl:2]: why was this deleted?

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl [hunk 21]
```diff
@@ -1,13 +1,49 @@
...
-
-# Fake values since non-FSAL
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl:4]: why was this deleted?

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl [hunk 22]
```diff
@@ -1072,20 +974,9 @@ function alg_cache(
...
-    du2 = zero(rate_prototype)
-    k1 = zero(rate_prototype)
-    k2 = zero(rate_prototype)
-    k3 = zero(rate_prototype)
-    k4 = zero(rate_prototype)
-    k5 = zero(rate_prototype)
-    k6 = zero(rate_prototype)
-    k7 = zero(rate_prototype)
-    k8 = zero(rate_prototype)
+    dense = zeros(eltype(rate_prototype), 3)
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl:977]:
```suggestion
    dense = [zero(rate_protoype) for i in 1:3]
```

otherwise it's not actually a zero vector.

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl [hunk 23]
```diff
@@ -1043,10 +946,9 @@ function alg_cache(alg::Rodas5, u, rate_prototype, ::Type{uEltypeNoUnits},
     linsolve = init(linprob, alg.linsolve, alias_A = true, alias_b = true,
         Pl = Pl, Pr = Pr,
         assumptions = LinearSolve.OperatorAssumptions(true))
-    grad_config = build_grad_config(alg, f, tf, du1, t)
-    jac_config = build_jac_config(alg, f, uf, du1, uprev, u, tmp, du2)
-    Rosenbrock5Cache(u, uprev, dense1, dense2, dense3, du, du1, du2, k1, k2, k3, k4,
-        k5, k6, k7, k8,
+    grad_config = build_grad_config(alg, f, tf, dus[2], t)
+    jac_config = build_jac_config(alg, f, uf, dus[2], uprev, u, tmp, dus[3])
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl:950]: why was this changed to 2 and 3?

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl [hunk 24]
```diff
@@ -39,7 +39,7 @@ end
 
     if repeat_step
         f(integrator.fsalfirst, uprev, p, t)
-        OrdinaryDiffEqCore.increment_nf!(integrator.stats, 1)
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_perform_step.jl:42]: why was this reverted? Please do not revert that, it will cause test failures.

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_tableaus.jl [hunk 25]
```diff
@@ -677,74 +503,35 @@ function Rodas5Tableau(T, T2)
...
 
-    Rodas5Tableau(a21, a31, a32, a41, a42, a43, a51, a52, a53, a54,
-        a61, a62, a63, a64, a65,
-        C21, C31, C32, C41, C42, C43, C51, C52, C53, C54,
-        C61, C62, C63, C64, C65, C71, C72, C73, C74, C75, C76,
-        C81, C82, C83, C84, C85, C86, C87,
-        gamma, d1, d2, d3, d4, d5, c2, c3, c4, c5,
-        h21, h22, h23, h24, h25, h26, h27, h28, h31, h32, h33, h34, h35, h36, h37,
-        h38, h41, h42, h43, h44, h45, h46, h47, h48)
+    Rodas5Tableau(a, C, gamma, d, c,
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_tableaus.jl:506]:
```suggestion
    RodasTableau(a, C, gamma, d, c,
```

?

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_tableaus.jl [hunk 26]
```diff
@@ -773,12 +560,9 @@ function Rodas5PTableau(T, T2)
     h47 = convert(T, -0.15582684282751913)
     h48 = convert(T, 4.883087185713722)
 
-    Rodas5Tableau(a21, a31, a32, a41, a42, a43, a51, a52, a53, a54,
-        a61, a62, a63, a64, a65,
-        C21, C31, C32, C41, C42, C43, C51, C52, C53, C54,
-        C61, C62, C63, C64, C65, C71, C72, C73, C74, C75, C76,
-        C81, C82, C83, C84, C85, C86, C87,
-        gamma, d1, d2, d3, d4, d5, c2, c3, c4, c5,
+    Rodas5Tableau(a,
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_tableaus.jl:563]:
```suggestion
    RodasTableau(a,
```

?

File: lib/OrdinaryDiffEqRosenbrock/src/stiff_addsteps.jl [hunk 27]
```diff
@@ -644,194 +547,51 @@ function _ode_addsteps!(k, t, uprev, u, dt, f, p, cache::Rosenbrock5Cache,
...
-                                h36 * k6 + h37 * k7 + h38 * k8
-        copyat_or_push!(k, 2, copy(tmp))
-
-        @.. broadcast=false tmp=h41 * k1 + h42 * k2 + h43 * k3 + h44 * k4 + h45 * k5 +
-                                h46 * k6 + h47 * k7 + h48 * k8
-        copyat_or_push!(k, 3, copy(tmp))
+        
+        # Calculate the final results for k
+        for j in 1:3
+            tmp = sum(eval(Symbol("h$j$i")) * ks[i] for i in 1:8)
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqRosenbrock/src/stiff_addsteps.jl:592]: no eval allowed. it should probably be an array anyways.

**Comment by ChrisRackauckas** [comment]
Lots of comments. You really need to fix the merge commit before trying to continue. You cannot just delete all changes that have happened, that's going to cause test failures. There is a naming overlap with one of the new abstract types, you'll want to actually handle that and not just revert it.

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl (hunk 23 above)
Comment (ParamThakkar123) [reply lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl:950]: actually dus had three variables du, du1, and du2 so I kept them in single vector as [du, du1, du2] according to Julia's 1-based indexing du1 would be dus[2] so I went that way

**Comment by ParamThakkar123** [comment]
> Lots of comments. You really need to fix the merge commit before trying to continue. You cannot just delete all changes that have happened, that's going to cause test failures. There is a naming overlap with one of the new abstract types, you'll want to actually handle that and not just revert it.

Okay. Will keep that in mind. working towards that. Sorry for the mistakes made

**Comment by ChrisRackauckas** [comment]
I highly recommend moving this PR to a branch so you can keep a clean master. This was causing you issues before, and now it's interfering with the other PR. Generally, don't open PRs from your fork's master.

File: lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl (hunk 21 above)
Comment (ParamThakkar123) [reply lib/OrdinaryDiffEqRosenbrock/src/rosenbrock_caches.jl]: Actually I wanted to club the constant caches into a single RosenbrockConstantCache and the names were conflicting. So I decided to keep one.
